### PR TITLE
Resolve types in parser

### PR DESCRIPTION
**Pull request by lenticularis39**
Instead of parsing types in casts as identifiers, implement separate tokens for type identifiers and parse type expressions. Simple types, parametric size types (like strings) and struct types are all supported. For types whose names collide with builtin functions the _t suffix is used (e.g. min_t).

The main purpose of this is preparation for typing arguments of user-defined functions (see #2516). Currently, it adds arbitrary pointer type nesting and sizeofs of arbitrary types; no new types are supported for casts.

Implementation notes:
- Struct types are generated with empty pointers in the parser and completed in SemanticAnalyser after the types are fetched by FieldAnalyser.
- sizeof is now a special kind of expression instead of being a call, since type arguments are now not treated as expressions.
- unit tests using invalid types had to be changed to use valid ones or removed, otherwise the code would be rejected by the parser

##### Checklist

- [ ] Language changes are updated in `man/adoc/bpftrace.adoc` and if needed in `docs/reference_guide.md`
- [ ] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

### CODE REVIEW

**Comment by danobi** [comment]
It's been a while since I've looked at the gnarly language bits so I don't have much to add. But it all looks pretty reasonable to me. All the tests passing is a good sign. :)

File: src/ast/ast.cpp [hunk 1]
```diff
@@ -321,18 +337,10 @@
...
-           bool is_pointer,
-           bool is_double_pointer,
-           Expression *expr,
-           location loc)
-    : Expression(loc),
-      cast_type(type),
-      is_pointer(is_pointer),
-      is_double_pointer(is_double_pointer),
-      expr(expr)
+Cast::Cast(SizedType cast_type, Expression *expr, location loc)
```
[the remark this answers is not in the document]
Comment (danobi) [reply src/ast/ast.cpp:340]: Probably bogus. A copy needs to be made at some point and I bet the compiler can elide extra copies here

**Review by danobi — APPROVED** [review]

**Comment by danobi** [comment]
@ajor @fbs @viktormalik any comments? Would be good to get this in soon-ish b/c this is a big change and will start generating merge conflicts soon

File: src/ast/ast.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (viktormalik) [reply src/ast/ast.cpp:340]: Agreed, dismissing.

**Review by viktormalik — APPROVED** [review]
Looks very reasonable, I like the thorough unit tests. Agreed with merging ASAP.

**Comment by ajor** [comment]
This change has broken some existing casting behaviour, which the removed tests "cast_typedef" and "cast_ptr_typedef" were covering.

It was desired behaviour for casts to be allowed without the "struct" prefix - see #682 for examples.

We should sort this out before building any more work on top of this change.

**Comment by lenticularis39** [comment]
Good point, thanks, I didn't know this is actually a feature of bpftrace, the documentation only mentions `struct xyz`.

The proper fix would be expanding at least `pointer_type` to also accept `IDENT "*"`, but that causes a reduce-reduce conflict somehow. Creating a nonterminal symbol for types and identifiers (i.e. typedef types) causes some reduce-reduce conflicts needing more investigation.

Until that is done, I suggest a workaround in a PR I'll link in a few minutes here (Since the fix described above is mostly need to use typedef types as an argument of a function, I'll probably include it in the PR introducing functions, unless it'll be need for something else.)

**Pull request merged**